packages/profile-sync-controller/src/controllers/user-storage/UserStorageController.ts

```diff
@@ -473,6 +473,53 @@ export default class UserStorageController extends BaseController<
     );
   }
 
+  #isLocked = false;
+
+  readonly #pendingOperations: (() => Promise<unknown> | unknown)[] = [];
+
+  /**
+   * Locks the controller, preventing operations that are subsequently called with `executeWithLockHandling` from executing.
+   * These operations will be queued until the controller is unlocked.
+   */
+  lock() {
+    this.#isLocked = true;
+  }
+
+  /**
+   * Unlocks the controller and executes all pending operations that were queued using the `executeWithLockHandling` method while the controller was locked.
+   * This method will also clear the pending operations queue.
+   * If any of the operations fail, the error will be propagated to the caller. The following operations will not be executed and the queue will be cleared.
+   */
+  async unlock() {
+    try {
+      for (const operation of this.#pendingOperations) {
+        await operation();
+      }
+    } finally {
+      this.#pendingOperations.length = 0;
+      this.#isLocked = false;
+    }
+  }
+
+  /**
+   * Executes or enqueues an operation based on the lock state of the controller.
+   * If the controller is locked, the operation will be added to the pending operations queue.
+   * If the controller is not locked, the operation will be executed immediately.
+   *
+   * @param operation - The operation to execute. This should be a function that returns a promise.
+   * @returns The result of the operation.
+   */
+  async executeWithLockHandling<T>(
+    operation: () => Promise<T>,
+  ): Promise<T | void> {
+    if (this.#isLocked) {
+      this.#pendingOperations.push(operation);
+      return;
+    }
+
+    await operation();
```

Comment:
There is a risk here, that pending operations are still in the queue when this new operation is executed, so this latest `await operation()` would take precedence.
We can't guarantee that operations are idempotent or commutative, so it's probably better to add new operations to the queue regardless of the locked state.

If we do this, we would need to ensure that new operations are consumed from the queue as well, since the queue itself is triggered by an `unlock()` but we might already be unlocked.
This means that we would have to trigger the consumption of operations here as well, if it is already unlocked.